proposals/v3/api.md

```diff
@@ -0,0 +1,815 @@
+# Synse v3 HTTP API
+
+
+Following the pattern established in previous versions of Synse Server,
+the URI prefix for requests will be `/synse/{API_VERSION}`, except for the two
+unversioned endpoints (`/test` and `/version`) where the URI prefix is just
+`/synse`.
+
+> **Highlights of Major Changes**:
+>  - `/scan` response change: removes device hierarchy, returns flattened list of devices
+>  - `/read` behavior change: no longer uses hierarchical routing, uses tags
+>  - `/write` behavior change: no longer uses hierarchical routing, uses tags
+>  - removes alias routes - behavior similar to alias routes can now be done via vanilla `/read` and `/write`
+
+
+> We could keep the alias routes around, but looking forward, it doesn't seem scalable since we could
+> have tens or hundreds of different device types supported by the system. In addition, not all of the
+> "LED" type devices may support the same thing, e.g., so it doesn't make sense to handle all LEDs the
+> same way. We could make a more generic means of checking inputs per device, not just per device type,
+> so it wouldn't matter how many device types there are and how they vary -- they could be checked automatically
+> and the API wouldn't need to be updated/modified for each of them.
+
+> Do we want to support more ways to access the API? Just HTTP? Add external gRPC api? websockets?
+> We probably don't need this configurability right now, but it could be worth considering or looking
+> in to, particularly if we start measuring system performance and need to improve it.
+
+## Errors
+Most errors returned from Synse Server will come back with a JSON payload in order to
+provide additional context for the error. Some errors will not return a JSON payload;
+this class of error is generally due to the service not being ready, available, or
+reachable.
+
+### Codes
+An error response will be returned with one of the following HTTP codes:
+
+* **400**: Invalid user input. This can range from invalid POSTed JSON, unsupported query
+  parameters being used, or invalid resource types.
+* **404**: Invalid URL or the specified resource is not found.
+* **500**: Server-side processing error.
+
+> **Question**: Should we support more error codes (finer grained)? In the past we've tried to use
+> as few codes as we can get away with to keep things simple. Does it make sense to return more
+> fine-grained error codes, or do these "high level" buckets suffice?
```

Comment:
No other error codes have occurred because we haven't encoded any other ones to occur. I think that finer grained error codes would give us more usable context/things, particularly around some of the async operations (see: https://farazdagi.com/2014/rest-and-long-running-jobs/ -- I read through this, which was the basis of me thinking about all of this in the first place). 

That said, it very much so could be overkill for us at this point, and the internal error code could be enough. Another thing that influenced this idea was the notion of having some sort of security/access (e.g. role based access to resources, which I believe has a brief writeup in a different doc), which would require additional error codes. Again, its not clear that we need that now, or that it would even be at the Synse Server level. 

I think you're right that keeping stuff in high-level buckets is the way to go right now, but I'm still not entirely sure.

Reply:
The reading makes sense. In the near future, when we have discussions about this, I would love to learn about it more.